node/lib/util/submodule_util.js

```diff
@@ -305,48 +305,80 @@ exports.getSubmoduleRepos = co.wrap(function *(repo) {
-        lastSubs = yield exports.getSubmodulesForCommit(repo, parentCommit);
+    // We calculate the changes of a commit against it's first parent.  If it
+    // has no parents, then the calculation is against an empty tree.
+
+    let parentTree = null;
+    const parents = yield commit.getParents();
+    if (0 !== parents.length) {
+        parentTree = yield parents[0].getTree();
     }
-    let result = {
-        added  : new Set(),
-        changed: new Set(),
-        removed: new Set(),
+
+    const tree = yield commit.getTree();
+    const diff = yield NodeGit.Diff.treeToTree(repo, parentTree, tree, null);
+    const num = diff.numDeltas();
+    const result = {
+        added: {},
+        changed: {},
+        removed: {},
+        modulesFileChanged: false,
     };
-    for (let name in currentSubs) {
-        const last = lastSubs[name];
-        if (undefined === last) {
-            result.added.add(name);
-        }
-        else {
-            const current = currentSubs[name];
-            if (current.sha !== last.sha || current.url !== last.url) {
-                result.changed.add(name);
-            }
-        }
-    }
-    for (let name in lastSubs) {
-        if (!(name in currentSubs)) {
-            result.removed.add(name);
+    const DELTA = NodeGit.Diff.DELTA;
+    const COMMIT = NodeGit.TreeEntry.FILEMODE.COMMIT;
+    for (let i = 0; i < num; ++i) {
+        const delta = diff.getDelta(i);
+        switch (delta.status()) {
+            case DELTA.MODIFIED:
+            case DELTA.CONFLICTED: {
+                const newFile = delta.newFile();
+                const path = newFile.path();
+                if (COMMIT === newFile.mode()) {
+                    result.changed[path] = {
+                        "new": newFile.id().tostrS(),
+                        "old": delta.oldFile().id().tostrS(),
+                    };
+                }
+                else if (SubmoduleConfigUtil.modulesFileName === path) {
+                    result.modulesFileChanged = true;
+                }
+            } break;
+            case DELTA.ADDED: {
+                const newFile = delta.newFile();
+                const path = newFile.path();
+                if (COMMIT === newFile.mode()) {
+                    result.added[newFile.path()] = newFile.id().tostrS();
+                }
+                else if (SubmoduleConfigUtil.modulesFileName === path) {
+                    result.modulesFileChanged = true;
+                }
+            } break;
+            case DELTA.DELETED: {
+                const oldFile = delta.oldFile();
+                const path = oldFile.path();
+                if (COMMIT === oldFile.mode()) {
+                    result.removed[oldFile.path()] = oldFile.id().tostrS();
+                }
+                else if (SubmoduleConfigUtil.modulesFileName === path) {
+                    result.modulesFileChanged = true;
+                }
+            } break;
```

Comment:
case DELTA.RENAMED.  case DELTA.COPIED.

Reply:
I put in a TODO for now. Not sure if these crop up with submodules or how I'd handle yet; I need to do more testing.

Reply:
IIRC, I have seen them while doing some manual gitlab testing at one point.  But I could be misremembering.  At the very least, we ought to die when we see them so that we don't just skip them.